docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+
+[web3.storage](http://web3.storage) is a Storefront providing APIs to enable users to easily upload CAR files, while getting them available on the IPFS Network and stored in multiple locations via Filecoin Storage Providers. It relies on Spade as a broker to get their user data into Filecoin Storage Providers. Currently, Spade requires a Filecoin Piece with size between 15.875GiB and 31.75GiB to create deals with Filecoin Storage Providers. Moreover, the closer a Filecoin Piece is closer to the upper bound, the most optimal are the associated storage costs.
+
+Taking into account that [web3.storage](http://web3.storage) onboards any type of content (up to a maximum of 4GiB-padded shards to have better utilization of Fil sector space), multiple CAR files uploaded need to be aggregated into a bigger Piece that can be offered to Filecoin Storage Providers. w3filecoin pipeline keeps track of queued CARs (cargo) to be included in Storage Provider deals.
+
+When a CAR file is written into a given web3.storage's bucket, its metadata gets into the w3filecoin processing pipeline. This pipeline is composed of multiple processing queues, together with a job scheduler per queue that will perform the processing. Each queue handles a processing stage with the goal of getting CAR files into Filecoin deals with Storage Providers.
+
+## High Level design
+
+The high level flow for the w3filecoin Pipeline is:
+
+- **Event** is triggered once a CAR file is written into a bucket with its metadata {`link`, `size`, `bucketName`, `bucketEndpoint`}. This event is added to a `content_validator_queue`.
+- **Content Validator process** validates CARs and writes references to a `content` table.
+- On its own schedule, **Piece maker process** can pull queued content from a `content_queue`, derive piece info for them and write records to a `piece` & `inclusion` tables.
+  - a `inclusion` table enables same piece to be included in another aggregate if a deal fails.
+- **Agregagtor process** reads from a `cargo_queue` (backed by `inclusion` table), attempts to create an aggregate and if successful it writes to an `aggregate` table.
+- **Submission process** reads from the `aggregate_queue`, submits aggregates to the agency (spade proxy) and writes deal record with status "PENDING".
+- TBD deal flow
+
+![Pipeline processes](./processes.png)
+
+The w3filecoin pipeline is modeled into 4 different SST Stacks that will have their infrastructure provisioned in AWS via AWS CloudFormation. These are:
+
+- API Stack
+- Consumer Stack
+- DB Stack
+- Processor Stack
+
+![Architecture](./architecture.png)
+
+## API Stack
+
+TBC
+API Gateway to expose:
+- Report API for failed aggregates to land into Storage Providers
+- Get to know state of deals
+- ...
+
+## Consumer Stack
+
+w3filecoin relies on events from Buckets once CAR files are written into them. It is designed to enable multiple sources of CAR files to be easily integrated into the pipeline via its consumer stack.
```

Comment:
I think it might be a good idea to decouple it from bucket events and simply have a mechanism for adding things into consumer pipeline. This way S3 buckets events could be one source of events, but we could also swap or introduce other event source e.g switch to R2 and use content claim invocation as a trigger.

Reply:
I will add this once we get the implementation part. Now dropping the details